sdk/pinocchio/src/sysvars/slot_hashes/mod.rs

```diff
@@ -0,0 +1,443 @@
+    ///
+    /// This constructor performs comprehensive validation of the data format
+    /// including length prefix, entry count bounds, and buffer size requirements.
+    /// Does not validate that entries are sorted in descending order.
+    #[inline(always)]
+    pub fn new(data: T) -> Result<Self, ProgramError> {
+        let len = parse_and_validate_data(&data)?;
+        Ok(unsafe { Self::new_unchecked(data, len) })
+    }
+
+    /// Creates a `SlotHashes` instance directly from a data container and entry count.
+    /// Important: provide a valid len. Whether or not len is assumed to be
+    /// the constant 20_488 (512 entries) is up to caller.
+    ///
+    /// # Safety
+    ///
+    /// This function is unsafe because it does not check the validity of the data or count.
+    /// The caller must ensure:
+    /// 1. The underlying byte slice in `data` represents valid SlotHashes data
+    ///    (length prefix + entries, where entries are sorted in descending order by slot).
+    /// 2. `len` is the correct number of entries (≤ MAX_ENTRIES), matching the prefix.
+    /// 3. The data slice contains at least `NUM_ENTRIES_SIZE + len * ENTRY_SIZE` bytes.
+    ///
+    #[inline(always)]
+    pub unsafe fn new_unchecked(data: T, len: usize) -> Self {
+        debug_assert!(len <= MAX_ENTRIES && data.len() >= NUM_ENTRIES_SIZE + len * ENTRY_SIZE);
+
+        SlotHashes { data, len }
+    }
+
+    /// Gets the number of entries stored in this SlotHashes instance.
+    /// Performs validation checks and returns the entry count if valid.
+    #[inline(always)]
+    pub fn get_entry_count(&self) -> Result<usize, ProgramError> {
+        let data_entry_count = read_entry_count_from_bytes(&self.data).unwrap_or(0);
+        if data_entry_count != self.len {
+            return Err(ProgramError::InvalidArgument);
+        }
+        Ok(self.len)
+    }
+
+    /// Reads the entry count directly from the beginning of this SlotHashes instance **without validation**.
+    ///
+    /// # Safety
+    ///
+    /// This function is unsafe because it performs no checks on the underlying data.
+    /// The caller **must** ensure that:
+    /// 1. The underlying data contains at least `NUM_ENTRIES_SIZE` (8) bytes.
+    /// 2. The first 8 bytes represent a valid `u64` in little-endian format.
+    /// 3. Calling this function without ensuring the above may lead to panics
+    ///    (out-of-bounds access) or incorrect results.
+    #[inline(always)]
+    pub unsafe fn get_entry_count_unchecked(&self) -> usize {
+        read_entry_count_from_bytes_unchecked(&self.data)
+    }
+
+    /// Validates offset parameters for fetching SlotHashes data.
+    ///
+    /// # Arguments
+    /// * `offset` - Byte offset within the sysvar data
+    /// * `buffer_len` - Length of the buffer that will receive the data
+    ///
+    /// # Returns
+    /// Ok(()) if the offset is valid, Err otherwise
+    #[inline(always)]
+    pub(crate) fn validate_fetch_offset(
```

Comment:
`validate_fetch_offeset()`, `fetch_into()`, and `fetch_into_unchecked()` seem to provide an interface to dealing with the slot hashes that is an alternative to using the `SlotHashes` type.

Would it make sense to move these functions into either a separate module or, at least, outside the `SlotHashes` implementation, to the top level of this module?

I think a clear separation could help users understand which parts of the API they care about and which parts they do not care about.
I would imagine each case will either use a `SlotHashes` instance with an account, or the "caller controlled buffer" API.  Most would not want to use both at the same time.

Reply:
Yes, this seems like a natural progression. Back in a bit.

Reply:
These now live in raw.rs